core/src/block.rs

```diff
@@ -95,74 +101,74 @@ impl Chain {
     }
 }
 
-/// Chain iterator
-pub struct ChainIterator<'itm> {
-    chain: &'itm Chain,
-    pos_front: u64,
-    pos_back: u64,
+/// Blockchain viewer.
+/// Keeps an internal lock on the blockchain in order
+/// for you to read it. Allows only read operations on
+/// the chain and will block everyone else from operating
+/// on the chain. Works across async and non async threads.
+///
+/// Try not to spam the lock, aka acquiring frequently, more
+/// than needed. If you want to keep long term access to information
+/// consider copying it out of the chain and dropping the
+/// `LockedChainViewer`
+pub struct LockedChainViewer<'chain> {
+    chain_mutex_guard: MutexGuard<'chain, Chain>,
 }
 
-impl<'itm> ChainIterator<'itm> {
-    fn new(chain: &'itm Chain) -> Self {
-        ChainIterator {
-            chain,
-            pos_front: 1,
-            pos_back: chain.len() as u64,
+impl<'chain> LockedChainViewer<'chain> {
+    /// Create new `LockedChainViewer` from `Chain` mutexguard.
+    pub fn new(chain_mutex_guard: MutexGuard<'chain, Chain>) -> Self {
+        Self { chain_mutex_guard }
+    }
+    /// Look for a block with the provided hash starting at the front
+    /// of the chain going backward. If it is found return a slice
+    /// of blocks starting with the one with the provided hash and
+    /// ending with the latest block.
+    #[allow(clippy::unwrap_in_result, clippy::expect_used, clippy::cast_sign_loss)]
+    pub fn blocks_after_hash(
+        &self,
+        hash: HashOf<VersionedCommittedBlock>,
+    ) -> Option<&[VersionedCommittedBlock]> {
+        if self.chain_mutex_guard.is_empty() {
+            return None;
         }
-    }
-
-    const fn is_exhausted(&self) -> bool {
-        self.pos_front > self.pos_back
-    }
-}
 
-impl<'itm> Iterator for ChainIterator<'itm> {
-    type Item = MapRef<'itm, u64, VersionedCommittedBlock>;
-    fn next(&mut self) -> Option<Self::Item> {
-        if !self.is_exhausted() {
-            let val = self.chain.blocks.get(&self.pos_front);
-            self.pos_front += 1;
-            return val;
+        let mut index: isize = (self.chain_mutex_guard.blocks.len() - 1)
+            .try_into()
+            .expect("we don't have more than 2^63 blocks");
+        while index >= 0 {
+            if self.chain_mutex_guard.blocks[index as usize]
+                .header()
+                .previous_block_hash
+                == hash
+            {
+                return Some(&self.chain_mutex_guard.blocks[index as usize..]);
+            }
+            index -= 1;
         }
         None
```

Comment:
```suggestion
self
    .chain_mutex_guard
    .blocks
    .iter()
    .rev()
    .position(|block| block.header().previous_block_hash == hash)
    .map(|index| &self.chain_mutex_guard.blocks[index..])
```

Reply:
very good suggestion

Reply:
Well okay turns out this was a lot more complicated to get working than what I originally wrote so I think it's best to keep that.